### PR TITLE
[SPARK-12230][ML] WeightedLeastSquares.fit() should handle division by zero properly if standard deviation of target variable is zero.

### PR DESCRIPTION
**Pull request by iyounus**
This fixes the behavior of WeightedLeastSquars.fit() when the standard deviation of the target variable is zero. If the fitIntercept is true, there is no need to train.

### CODE REVIEW

File: mllib/src/main/scala/org/apache/spark/ml/optim/WeightedLeastSquares.scala [hunk 1]
```diff
@@ -94,7 +110,7 @@ private[ml] class WeightedLeastSquares(
       if (standardizeFeatures) {
         lambda *= aVar(j - 2)
       }
-      if (standardizeLabel) {
+      if (standardizeLabel && bStd != 0) {
         // TODO: handle the case when bStd = 0
```
Comment (sethah) [inline mllib/src/main/scala/org/apache/spark/ml/optim/WeightedLeastSquares.scala]: If this PR is meant to address this "TODO", then the comment should be removed.

File: mllib/src/main/scala/org/apache/spark/ml/optim/WeightedLeastSquares.scala [hunk 2]
```diff
@@ -86,6 +86,22 @@ private[ml] class WeightedLeastSquares(
...
+          s"zeros and the intercept will be the mean of the label; as a result, " +
+          s"training is not needed.")
+        val coefficients = new DenseVector(Array.ofDim(k-1))
+        val intercept = bBar
+        val diagInvAtWA = new DenseVector(Array.ofDim(k))
+        return new WeightedLeastSquaresModel(coefficients, intercept, diagInvAtWA)
+      }
+      else {
+      logWarning(s"The standard deviation of the label is zero. " +
+        "Consider setting FitIntercept=true.")
```
Comment (sethah) [inline mllib/src/main/scala/org/apache/spark/ml/optim/WeightedLeastSquares.scala]: nit: "FitIntercept=true" -> "fitIntercept = true"

File: mllib/src/test/scala/org/apache/spark/ml/optim/WeightedLeastSquaresSuite.scala [hunk 3]
```diff
@@ -43,6 +44,18 @@ class WeightedLeastSquaresSuite extends SparkFunSuite with MLlibTestSparkContext
       Instance(23.0, 3.0, Vectors.dense(2.0, 11.0)),
       Instance(29.0, 4.0, Vectors.dense(3.0, 13.0))
     ), 2)
+
+    /*
+       R code:
+       same as above except make the label constant
```
Comment (sethah) [inline mllib/src/test/scala/org/apache/spark/ml/optim/WeightedLeastSquaresSuite.scala]: IMO, comments referencing things like "above" should be avoided since the code can be rearranged. This comment would be confusing if the above data is declared in a different place or removed altogether. I don't think think pointing out that it is the same as the other data is necessary here.

File: mllib/src/main/scala/org/apache/spark/ml/optim/WeightedLeastSquares.scala [hunk 4]
```diff
@@ -86,6 +86,22 @@ private[ml] class WeightedLeastSquares(
...
     val aaValues = aaBar.values
 
+    if (bStd == 0) {
+      if (fitIntercept) {
+        logWarning(s"The standard deviation of the label is zero, so the coefficients will be " +
+          s"zeros and the intercept will be the mean of the label; as a result, " +
+          s"training is not needed.")
+        val coefficients = new DenseVector(Array.ofDim(k-1))
+        val intercept = bBar
+        val diagInvAtWA = new DenseVector(Array.ofDim(k))
```
Comment (sethah) [inline mllib/src/main/scala/org/apache/spark/ml/optim/WeightedLeastSquares.scala]: Looking at the `LinearRegression` class, it handles the case where `bStd` == 0 by returning `DenseVector(Array(0D))` for `diagInvAtWA`. Then, in the training summary there are several cases to catch this invalid value for `diagInvAtWA`. This should be changed to:

```scala
val diagInvAtWA = new DenseVector(Array(0D))
```

**Comment by jkbradley** [comment]
@dbtsai Would you have time to take a look at this?  Thank you!

**Comment by dbtsai** [comment]
jenkins, add to whitelist

File: mllib/src/main/scala/org/apache/spark/ml/optim/WeightedLeastSquares.scala [hunk 5]
```diff
@@ -86,6 +86,22 @@ private[ml] class WeightedLeastSquares(
     val aaBar = summary.aaBar
     val aaValues = aaBar.values
 
+    if (bStd == 0) {
+      if (fitIntercept) {
```
Comment (dbtsai) [inline mllib/src/main/scala/org/apache/spark/ml/optim/WeightedLeastSquares.scala]: The `LinearRegression` has a bug related to this,

https://github.com/apache/spark/blob/master/mllib/src/main/scala/org/apache/spark/ml/regression/LinearRegression.scala#L226

when fitIntercept is false, the code should still train the model. Can you fix it in either separate PR or here? 

Thanks.

File: mllib/src/main/scala/org/apache/spark/ml/optim/WeightedLeastSquares.scala [hunk 6]
```diff
@@ -86,6 +86,22 @@ private[ml] class WeightedLeastSquares(
...
+        logWarning(s"The standard deviation of the label is zero, so the coefficients will be " +
+          s"zeros and the intercept will be the mean of the label; as a result, " +
+          s"training is not needed.")
+        val coefficients = new DenseVector(Array.ofDim(k-1))
+        val intercept = bBar
+        val diagInvAtWA = new DenseVector(Array(0D))
+        return new WeightedLeastSquaresModel(coefficients, intercept, diagInvAtWA)
+      }
+      else {
```
Comment (dbtsai) [inline mllib/src/main/scala/org/apache/spark/ml/optim/WeightedLeastSquares.scala]: move `else` up.

File: mllib/src/main/scala/org/apache/spark/ml/optim/WeightedLeastSquares.scala (hunk 5 above)
Comment (dbtsai) [inline mllib/src/main/scala/org/apache/spark/ml/optim/WeightedLeastSquares.scala]: Let's fix it in a separate PR to make thing easier.

File: mllib/src/main/scala/org/apache/spark/ml/optim/WeightedLeastSquares.scala [hunk 7]
```diff
@@ -94,8 +110,7 @@ private[ml] class WeightedLeastSquares(
       if (standardizeFeatures) {
         lambda *= aVar(j - 2)
       }
-      if (standardizeLabel) {
-        // TODO: handle the case when bStd = 0
+      if (standardizeLabel && bStd != 0) {
```
Comment (dbtsai) [inline mllib/src/main/scala/org/apache/spark/ml/optim/WeightedLeastSquares.scala]: Can you check when `standardizeLabel = true` and `bStd == 0.0`, what is the solution from R, and add it into unit-test?

File: mllib/src/main/scala/org/apache/spark/ml/optim/WeightedLeastSquares.scala (hunk 5 above)
Comment (iyounus) [inline mllib/src/main/scala/org/apache/spark/ml/optim/WeightedLeastSquares.scala]: I did notice that bug. I was planning to create separate jira for this.

File: mllib/src/main/scala/org/apache/spark/ml/optim/WeightedLeastSquares.scala (hunk 7 above)
Comment (iyounus) [inline mllib/src/main/scala/org/apache/spark/ml/optim/WeightedLeastSquares.scala]: @dbtsai The problem here is that for regularized regression in R, I need to use `glmnet`. But for this specific case (constant label, no intercept and no regularization) the results from `glmnet` do no match with `lm`. So I see a discrepancy within R itself. Have a look at the following R code:

```
A <- matrix(c(0, 1, 2, 3, 5, 7, 11, 13), 4, 2)  
b <- c(17, 17, 17, 17)  
w <- c(1, 2, 3, 4)  
df <- as.data.frame(cbind(A, b))

lm.model <- lm(b ~ . -1, data=df, weights=w)
print(as.vector(coef(lm.model)))
[1] -9.221298  3.394343

glm.model <- glmnet(A, b, weights=w, intercept=FALSE, lambda=0,
                    standardize=FALSE, alpha=0, thresh=1E-14)
print(as.vector(coef(glm.model)))
[1] 0 0 0
```

Note that in this example, I expect same results from both `lm` and `glmnet` because I've set `lambda=0` in `glmnet`. (BTW `standardize` has not effect here.) It seems to me that `glmnet` just sets all coefficients to zero if label is constant and intercept is not included. This is true even if I include regularization.

Right now `WeightedLeastSquares` (without regularization) matches with `lm`, and I think this is the correct behaviour given my understanding of the normal equation. With regularization, it should still give some non-zero coefficients, which is does. I don't know why `glmnet` behaves differently, but I don't think we should try to match that in this particular case.

File: mllib/src/main/scala/org/apache/spark/ml/optim/WeightedLeastSquares.scala (hunk 5 above)
Comment (iyounus) [inline mllib/src/main/scala/org/apache/spark/ml/optim/WeightedLeastSquares.scala]: @dbtsai I just created PR for this bug.

File: mllib/src/main/scala/org/apache/spark/ml/optim/WeightedLeastSquares.scala (hunk 7 above)
Comment (dbtsai) [inline mllib/src/main/scala/org/apache/spark/ml/optim/WeightedLeastSquares.scala]: Thanks. As you said, we will expect non zero coefficients in this case, so we don't have to match glmnet. 

However, we may want to throw excpetion when standerizeLabe is true, and ystd is zero since the problem is not well defined. 

Thanks.

File: mllib/src/test/scala/org/apache/spark/ml/optim/WeightedLeastSquaresSuite.scala [hunk 8]
```diff
@@ -74,6 +89,35 @@ class WeightedLeastSquaresSuite extends SparkFunSuite with MLlibTestSparkContext
...
+
+    val expected = Seq(
+      Vectors.dense(0.0, -9.221298, 3.394343),
+      Vectors.dense(17.0, 0.0, 0.0))
+
+    var idx = 0
+    for (fitIntercept <- Seq(false, true)) {
+      val wls = new WeightedLeastSquares(
+        fitIntercept, regParam = 0.0, standardizeFeatures = false, standardizeLabel = true)
+        .fit(instancesConstLabel)
```
Comment (dbtsai) [inline mllib/src/test/scala/org/apache/spark/ml/optim/WeightedLeastSquaresSuite.scala]: Added `standardizeLabel = false` with non-zero `regParam` with analytic normal equation solution. 

Added failure test of `standardizeLabel = true`, `regParam != 0` and `yStd == 0.0`.

File: mllib/src/main/scala/org/apache/spark/ml/optim/WeightedLeastSquares.scala (hunk 7 above)
Comment (iyounus) [inline mllib/src/main/scala/org/apache/spark/ml/optim/WeightedLeastSquares.scala]: The `WeightedLeastSquares` class is private and its instantiated in `LinearRegression` class where `standerizeLabe` parameter is hard wired to be `true`. So, the user doesn't have any control on this parameter.

We can throw an exception when `yStd` is zero and `regParam` is non-zero. But, if that is the case, then, why not to throw exception when `yStd` is zero regardless of what other parameters are? I cannot think of any interpretation of the model in this case.

The option could be to simply log a warning when we don't standardize the label here.

Let me know what you think.

Comment (dbtsai) [inline mllib/src/main/scala/org/apache/spark/ml/optim/WeightedLeastSquares.scala]: It's interesting to see that when regularization is zero, with/without standardization on labels or features will not change the solution of Linear Regression which you can experiment. 

As a result, the only issue that the model will be non-interpretable will be `yStd` is zeo and `regParam` is non-zero. You can have a `require` there with proper message. I think logging a warning will be probably very easy for users to ignore. Thanks.

Comment (dbtsai) [inline mllib/src/main/scala/org/apache/spark/ml/optim/WeightedLeastSquares.scala]: Here is the exercise,

```scala
  test("WLS against lm") {
    /*
       R code:

       df <- as.data.frame(cbind(A, b))
       for (formula in c(b ~ . -1, b ~ .)) {
         model <- lm(formula, data=df, weights=w)
         print(as.vector(coef(model)))
       }

       [1] -3.727121  3.009983
       [1] 18.08  6.08 -0.60
     */

    val expected = Seq(
      Vectors.dense(0.0, -3.727121, 3.009983),
      Vectors.dense(18.08, 6.08, -0.60))

    var idx = 0
    for (fitIntercept <- Seq(false, true)) {
      for (standardization <- Seq(false, true)) {
        val wls = new WeightedLeastSquares(
          fitIntercept, regParam = 0.0, standardizeFeatures = standardization,
          standardizeLabel = standardization).fit(instances)
        val actual = Vectors.dense(wls.intercept, wls.coefficients(0), wls.coefficients(1))
        assert(actual ~== expected(idx) absTol 1e-4)
      }
      idx += 1
    }
  }
```

File: mllib/src/test/scala/org/apache/spark/ml/optim/WeightedLeastSquaresSuite.scala (hunk 8 above)
Comment (iyounus) [inline mllib/src/test/scala/org/apache/spark/ml/optim/WeightedLeastSquaresSuite.scala]: I've added exception and the test for `standardizeLabel = true`, `regParam != 0` and `yStd == 0.0`. The only thing now left is to add test for the case `standardizeLabel = false`, `regParam != 0` and `yStd == 0.0`. As I mentioned before, in this case I cannot compare with `glmnet`. So, I'll try to implement normal equation in python by myself and compare with that. The good thing is that, for this  particular case, both `normal` and `l-bfgs` give identical results!

Comment (dbtsai) [inline mllib/src/test/scala/org/apache/spark/ml/optim/WeightedLeastSquaresSuite.scala]: Awesome! Great that you see the same result! For normal equation, will be nice to in R so we can have it in comment consistently. I did implement it once when I implemented the LBFGS version, let me try to find it.

**Comment by dbtsai** [comment]
LGTM. Wait for one extra test. Thanks.

File: mllib/src/test/scala/org/apache/spark/ml/optim/WeightedLeastSquaresSuite.scala (hunk 8 above)
Comment (iyounus) [inline mllib/src/test/scala/org/apache/spark/ml/optim/WeightedLeastSquaresSuite.scala]: @dbtsai I've implemented normal equation with regularization in R. Here is my code:
```
ridge_regression <- function(A, b, lambda, intercept=TRUE){
    if (intercept) {
        A = cbind(rep(1.0, length(b)), A)
        I = diag(ncol(A))
        I[1,1] = 0.0
    } else {
        I = diag(ncol(A))
    }
    R = chol( t(A) %*% A + lambda*I )
    z = solve(t(R), t(A) %*% b)
    w = solve(R, z)
    return(w)
}
```
And here are the results I get using this function.
```
A <- matrix(c(0, 1, 2, 3, 5, 7, 11, 13), 4, 2)
b <- c(17, 19, 23, 29)
ridge_regression(A, b, 0.1)
           [,1]
[1,] 12.9048179
[2,]  2.1151586
[3,]  0.6580494
```
The problem is that these don't quite match with `glmnet`. The difference can be at few percent level:
```
> model <- glmnet(A, b, intercept=TRUE, lambda=0.1, standardize=FALSE,
+ alpha=0, thresh=1E-20)
> print(as.vector(coef(model)))
[1] 13.1018870  2.2362361  0.6159732
```
But, my results match exactly with what I get from ridge regression in sklearn:

```
from sklearn.linear_model import Ridge
import numpy as np

A = np.array([[0, 1, 2, 3],[5, 7, 11, 13]]).T
b = np.array([17.0, 19.0, 23.0, 29.0])
model = Ridge(alpha=0.1, solver='cholesky', fit_intercept=True)
model.fit(A, b)
print model.intercept_
print model.coef_

12.9048178613
[ 2.11515864  0.65804935]
```
Even if I use other solvers (`svd`, `lsqr`, `sparse_cg`) in `sklearn.linear_model.Ridge`, I get exactly the same results.

If I don't use regularization by setting `lambda=0`, then the results from `glmnet` are identical to what I get from normal equation and `sklearn.linear_model.Ridge`.

Have you observed such differences before? Is `glmnet` making some other corrections or its just numerical precision issue?  I can't seem to reproduce `glmnet` results.

**Comment by mengxr** [comment]
Merged into master. Thanks!

File: mllib/src/test/scala/org/apache/spark/ml/optim/WeightedLeastSquaresSuite.scala (hunk 8 above)
Comment (dbtsai) [inline mllib/src/test/scala/org/apache/spark/ml/optim/WeightedLeastSquaresSuite.scala]: Sorry for getting you back so late. The difference is due to that `glmnet` always standardizes labels even `standardization == false`. `standardization == false` is turning off the standardization on features. As a result, at least in `glmnet`, when `ystd == 0.0`, the training is not valid.
